File: CSharp/Math/Matrix.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Reflection.Metadata.Ecma335;
using System.Text;

namespace CSharp
{
    public class Matrix
    {
        private readonly double[,] m;

        public Matrix() : this(4, 4)
        {
            // Nothing here
        }

        public Matrix(int rows, int columns) : this(rows, columns, null)
        {
            // Nothing here
        }

        public Matrix(int rows, int columns, double[] data)
        {
            m = new double[rows, columns];
            if (data != null)
            {
                LoadMatrix(rows, columns, data);
            }
        }

        public Matrix(Matrix matrix)
        {
            m = new double[matrix.Rows, matrix.Columns];

            for (int i = 0; i < matrix.Rows; i++)
                for (int j = 0; j < matrix.Columns; j++)
                    m[i, j] = matrix[i, j];
        }

        public Matrix(Tuple t, bool isColumn = true)
        {
            int rows = 1;
            int columns = 1;

            if (isColumn)
                rows = 4;
            else
                columns = 4;

            double[] data = { t.x, t.y, t.z, t.w };

            m = new double[rows, columns];
            LoadMatrix(rows, columns, data);
        }

        public int Rows => m.GetUpperBound(0) + 1;
        public int Columns => m.GetUpperBound(1) + 1;

        public double this[int row, int column]
        {
            get
            {
                return m[row, column];
            }

            set
            {
                m[row, column] = value;
            }
        }

        public static Matrix operator *(Matrix lhs, Matrix rhs)
        {
            if (lhs.Columns != rhs.Rows)
                throw new Exception("Matrix Cannot Be Multiplied.");

            Matrix ret = new Matrix(lhs.Rows, rhs.Columns);
            double temp = 0.0;

            for (int i = 0; i < lhs.Rows; i++)
            {
                for (int j = 0; j < rhs.Columns; j++)
                {
                    temp = 0;
                    for (int k = 0; k < lhs.Columns; k++)
                        temp += lhs[i, k] * rhs[k, j];
                    ret[i, j] = temp;
                }
            }

            return ret;
        }

        public static Tuple operator *(Matrix lhs, Tuple rhs)
        {
            if ((lhs.Rows != 4) || (lhs.Columns != 4))
                throw new Exception("Matrix must be Nx4 or 4xM.");

            Matrix value = lhs * new Matrix(rhs, lhs.Rows == 4);

            double x, y, z, w;
            if (value.Rows == 4)
            {
                x = value[0, 0];
                y = value[1, 0];
                z = value[2, 0];
                w = value[3, 0];
            }
            else
            {
                x = value[0, 0];
                y = value[0, 1];
                z = value[0, 2];
                w = value[0, 3];
            }

            return new Tuple(x, y, z, w);
        }

        public static Point operator *(Matrix lhs, Point rhs)
        {
            Tuple other = (Tuple)rhs;
            Tuple t = lhs * other;

            return new Point(t.x, t.y, t.z);
        }

        public static Vector operator *(Matrix lhs, Vector rhs)
        {
            Tuple other = (Tuple)rhs;
            Tuple t = lhs * other;

            return new Vector(t.x, t.y, t.z);
        }

        public static Matrix Identity
        {
            get
            {
                var m = new Matrix(4, 4);
                m[0, 0] = 1;
                m[1, 1] = 1;
                m[2, 2] = 1;
                m[3, 3] = 1;

                return m;
            }
        }

        public Matrix Transpose
        {
            get
            {
                Matrix m = new Matrix(Columns, Rows);

                for (int i = 0; i < Rows; i++)
                    for (int j = 0; j < Columns; j++)
                        m[j, i] = this[i, j];

                return m;
            }
        }

        public double Determinant
        {
            get
            {
                double det = 0;
                int size = Math.Max(Rows, Columns);
                if (size == 2)
                {
                    det = this[0, 0] * this[1, 1] - this[0, 1] * this[1, 0];
                }
                else
                {
                    for (var c = 0; c < size; c++)
                    {
                        det += this[0, c] * Cofactor(0, c);
                    }
                }

                return det;
            }
        }

        public Matrix Submatrix(int deleteRow, int deleteColumn)
        {
            Matrix m = new Matrix(Rows - 1, Columns - 1);
            int r = 0;
            int c = 0;

            for (int i = 0; i < Rows; i++)
            {
                if (i == deleteRow) continue;

                c = 0;
                for (int j = 0; j < Columns; j++)
                {
                    if (j == deleteColumn) continue;

                    m[r, c] = this[i, j];
                    c += 1;
                }

                r += 1;
            }

            return m;
        }

        public double Minor(int deleteRow, int deleteColumn)
        {
            Matrix m = Submatrix(deleteRow, deleteColumn);
            return m.Determinant;
        }

        public double Cofactor(int deleteRow, int deleteColumn)
        {
            Matrix m = Submatrix(deleteRow, deleteColumn);
            var det = m.Determinant;
            return (deleteRow + deleteColumn) % 2 == 0 ? det : -det;
        }

        public bool IsInvertible => !Determinant.IsEqual(0);

        public Matrix Inverse
        {
            get
            {
                if (!IsInvertible) throw new Exception("Matrix cannot be inverted.");

                Matrix m = new Matrix(Rows, Columns);
                var size = Math.Max(Rows, Columns);
                var det = Determinant;

                for (var row = 0; row < size; row++)
                {
                    for (var col = 0; col < size; col++)
                    {
                        var c = Cofactor(row, col);
                        m[col, row] = c / det;
                    }
                }

                return m;
            }
        }

        public static Matrix Translation(double x, double y, double z)
        {
            var id = Identity;
            id[0, 3] = x;
            id[1, 3] = y;
            id[2, 3] = z;

            return id;
        }

        public static Matrix Scaling(double x, double y, double z)
        {
            var id = Identity;
            id[0, 0] = x;
            id[1, 1] = y;
            id[2, 2] = z;

            return id;
        }

        public static Matrix RotationX(double angle)
        {
            var id = Identity;
            var sine = Math.Sin(angle);
            var cose = Math.Cos(angle);

            id[1, 1] = cose;
            id[1, 2] = -sine;
            id[2, 1] = sine;
            id[2, 2] = cose;

            return id;
        }

        public static Matrix RotationY(double angle)
        {
            var id = Identity;
            var sine = Math.Sin(angle);
            var cose = Math.Cos(angle);

            id[0, 0] = cose;
            id[0, 2] = sine;
            id[2, 0] = -sine;
            id[2, 2] = cose;

            return id;
        }

        public static Matrix RotationZ(double angle)
        {
            var id = Identity;
            var sine = Math.Sin(angle);
            var cose = Math.Cos(angle);

            id[0, 0] = cose;
            id[0, 1] = -sine;
            id[1, 0] = sine;
            id[1, 1] = cose;

            return id;
        }

        public static Matrix Shearing(double xy, double xz, double yx, double yz, double zx, double zy)
        {
            var id = Identity;

            id[0, 1] = xy;
            id[0, 2] = xz;
            id[1, 0] = yx;
            id[1, 2] = yz;
            id[2, 0] = zx;
            id[2, 1] = zy;

            return id;
        }

        public static Matrix ViewTransform(Point from, Point to, Vector up)
        {
            var forward = (new Vector(to - from)).Normalize;
            var upn = up.Normalize;
            var left = forward.Cross(upn);
            var trueUp = left.Cross(forward);
            var orientation = new Matrix(4, 4,
                                         new double[]
                                         {
                                             left.x,      left.y,     left.z,    0,
                                             trueUp.x,    trueUp.y,   trueUp.z,  0,
                                            -forward.x,  -forward.y, -forward.z, 0,
                                             0,           0,          0,         1
                                         });

            return orientation * Matrix.Translation(-from.x, -from.y, -from.z);
        }

        public override bool Equals(object obj)
        {
            var other = obj as Matrix;
            if ((other == null) ||
                (Rows != other.Rows) ||
                (Columns != other.Columns))
                return false;

            for (int i = 0; i < Rows; i++)
                for (int j = 0; j < Columns; j++)
                    if (!m[i, j].IsEqual(other[i, j]))
                        return false;

            return true;
        }

        public override int GetHashCode()
        {
            return m.GetHashCode();
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append($"{Rows}x{Columns} [");

            for (int i = 0; i < Rows; i++)
            {
                sb.Append("[");

                for (int j = 0; j < Columns; j++)
                {
                    sb.Append($"{m[i, j]} ");
                }

                sb.Append("] ");
            }
            sb.Append("]");

            return sb.ToString();
        }

        private void LoadMatrix(int rows, int columns, double[] data)
        {
            int index = 0;

            for (int i = 0; i < rows; i++)
                for (int j = 0; j < columns; j++)
                    m[i, j] = (data != null) && (index < data.Length) ? data[index++] : 0;
        }
    }
}

File: Java/src/main/java/Math/Matrix.java
package RTC.Math;

import java.util.Objects;

public class Matrix {
    public static int MatrixId = 0;

    private int id;
    private double[][] m;
    private int width;
    private int height;

    public Matrix(int width, int height) {
        this(width, height, new double[] { });
    }

    public Matrix(int width, int height, double[] array) {
        id = MatrixId++;
        this.width = width;
        this.height = height;

        m = new double[width][height];

        int index = 0;
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                m[x][y] = array.length > 0 ? array[index++] : 0;
            }
        }
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public void set(int x, int y, double value) {
        m[x][y] = value;
    }
    public double valueAt(int x, int y) {
        return m[x][y];
    }

    public Matrix mul(Matrix b) {
        Matrix M = new Matrix(width, height);

        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                M.set(x, y, m[x][0] * b.valueAt(0, y) +
                            m[x][1] * b.valueAt(1, y) +
                            m[x][2] * b.valueAt(2, y) +
                            m[x][3] * b.valueAt(3, y));
            }
        }

        return M;
    }

    public Tuple mul(Tuple b) {
        double x = b.getX();
        double y = b.getY();
        double z = b.getZ();
        double w = b.getW();

        double X = m[0][0] * x +
                   m[0][1] * y +
                   m[0][2] * z +
                   m[0][3] * w;
        double Y = m[1][0] * x +
                   m[1][1] * y +
                   m[1][2] * z +
                   m[1][3] * w;
        double Z = m[2][0] * x +
                   m[2][1] * y +
                   m[2][2] * z +
                   m[2][3] * w;
        double W = m[3][0] * x +
                   m[3][1] * y +
                   m[3][2] * z +
                   m[3][3] * w;

        return new Tuple(X, Y, Z, W);
    }

    public Matrix transpose() {
        Matrix M = new Matrix(height, width);

        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                M.set(y, x, m[x][y]);
            }
        }

        return M;
    }

    public double determinate() {
        double det = 0.0;

        if ((width == 2) && (height == 2)) {
            det = m[0][0] * m[1][1] - m[0][1] * m[1][0];
        } else {
            for (int c = 0; c < height; c++) {
                det = det + m[0][c] * cofactor(0, c);
            }
        }

        return det;
    }

    public Matrix submatrix(int row, int column) {
        Matrix M = new Matrix(width - 1, height - 1);

        int i = -1, j;
        for (int x = 0; x < width; x++) {
            if (x == row) continue;
            i++;

            j = -1;
            for (int y = 0; y < height; y++) {
                if (y == column) continue;
                j++;

                M.set(i, j, m[x][y]);
            }
        }

        return M;
    }

    public double minor(int row, int column) {
        Matrix M = submatrix(row, column);
        return M.determinate();
    }

    public double cofactor(int row, int column) {
        double m = minor(row, column);
        return (row + column) % 2 == 0 ? m : -m;
    }

    public boolean isInvertible() {
        return !Helper.IsEqual(determinate(), 0.0);
    }

    public Matrix inverse() {
        Matrix M = new Matrix(width, height);

        double c = 0.0;
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                c = cofactor(x, y);
                M.set(y, x, c / determinate());
            }
        }

        return M;
    }

    public static Matrix Identity() {
        return new Matrix(4, 4, new double[] { 1, 0, 0, 0,
                                               0, 1, 0, 0,
                                               0, 0, 1, 0,
                                               0, 0, 0, 1});
    }

    public static Matrix Translation(double x, double y, double z) {
        Matrix M = Identity();
        M.set(0, 3, x);
        M.set(1, 3, y);
        M.set(2, 3, z);

        return M;
    }

    public static Matrix Scaling(double x, double y, double z) {
        Matrix M = Identity();
        M.set(0, 0, x);
        M.set(1, 1, y);
        M.set(2, 2, z);

        return M;
    }

    public static Matrix RotationX(double r) {
        Matrix M = Identity();
        double cos = Math.cos(r);
        double sin = Math.sin(r);

        M.set(1, 1, cos);
        M.set(1, 2, -sin);
        M.set(2, 1, sin);
        M.set(2, 2, cos); 

        return M;
    }

    public static Matrix RotationY(double r) {
        Matrix M = Identity();
        double cos = Math.cos(r);
        double sin = Math.sin(r);

        M.set(0, 0, cos);
        M.set(0, 2, sin);
        M.set(2, 0, -sin);
        M.set(2, 2, cos); 

        return M;
    }

    public static Matrix RotationZ(double r) {
        Matrix M = Identity();
        double cos = Math.cos(r);
        double sin = Math.sin(r);

        M.set(0, 0, cos);
        M.set(0, 1, -sin);
        M.set(1, 0, sin);
        M.set(1, 1, cos); 

        return M;
    }

    public static Matrix Shearing(double xy, double xz, double yx, double yz, double zx, double zy) {
        Matrix M = Identity();

        M.set(0, 1, xy);
        M.set(0, 2, xz);
        M.set(1, 0, yx);
        M.set(1, 2, yz);
        M.set(2, 0, zx);
        M.set(2, 1, zy);

        return M;
    }

    public static Matrix ViewTransform(Point from, Point to, Vector up)
    {
        Tuple forward = to.sub(from).norm();
        Tuple upn = up.norm();
        Tuple left = forward.cross(upn);
        Tuple trueUp = left.cross(forward);
        Matrix orientation = new Matrix(4, 4,
                                        new double[]
                                        {
                                            left.getX(),     left.getY(),     left.getZ(),     0,
                                            trueUp.getX(),   trueUp.getY(),   trueUp.getZ(),   0,
                                            -forward.getX(), -forward.getY(), -forward.getZ(), 0,
                                            0,               0,               0,               1
                                        });

        return orientation.mul(Matrix.Translation(-from.getX(), -from.getY(), -from.getZ()));
    }    

    @Override
    public boolean equals(Object o)
    {
        if (o == this) {
            return true;
        }

        if ((o == null) || !(o instanceof Matrix)) {
            return false;
        }

        Matrix t = (Matrix) o;
        if ((width != t.getWidth()) || (height != t.getHeight())) {
            return false;
        }

        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                if (!Helper.IsEqual(m[x][y], t.valueAt(x, y))) {
                    return false;
                }
            }
        }

        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, m.hashCode());
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(id).append(" [");

        for (int x = 0; x < width; x++)
        {
            sb.append("[ ");
            for (int y = 0; y < height; y++)
            {
                sb.append(m[x][y]).append(" ");
            }
            sb.append("]");
        }

        sb.append("]");
        return sb.toString();
    }
}
